File: specsy-core/src/main/scala/org/specsy/core/Context.scala
// Copyright © 2010-2012, Esko Luontola <www.orfjackal.net>
// This software is released under the Apache License 2.0.
// The license text is at http://www.apache.org/licenses/LICENSE-2.0

package org.specsy.core

import Context._
import scala.collection.mutable.Buffer
import fi.jumi.api.drivers.{TestNotifier, SuiteNotifier}

object Context {
  abstract sealed class Status
  case object NotStarted extends Status
  case object Running extends Status
  case object Finished extends Status
}

class Context(targetPath: Path, notifier: SuiteNotifier) {
  private var status: Status = NotStarted
  private var current: SpecDeclaration = null
  private val postponed = Buffer[Path]()

  def bootstrap(className: String, rootSpec: => Unit) {
    changeStatus(NotStarted, Running)

    enterRootSpec(className)
    processSpec(rootSpec)
    exitSpec()

    changeStatus(Running, Finished)
  }

  private def enterRootSpec(name: String) {
    current = new SpecDeclaration(name, null, Path.ROOT, targetPath)
  }

  def specify(name: String, body: => Unit) {
    assertStatusIs(Running)

    enterSpec(name)
    processSpec(body)
    exitSpec()
  }

  private def enterSpec(name: String) {
    current = current.nextChild(name)
  }

  private def processSpec(body: => Unit) {
    if (current.shouldExecute) {
      execute(body)
    }
    if (current.shouldPostpone) {
      postponed += current.path
    }
  }

  private def execute(body: => Unit) {
    notifier.fireTestFound(current.path.toTestId, current.name)
    val tn = notifier.fireTestStarted(current.path.toTestId)

    executeSafely(new Closure {
      def run() {
        body
      }
    }, tn)
    import scala.collection.JavaConversions._
    current.deferred.foreach(executeSafely(_, tn))

    tn.fireTestFinished()
  }

  private def executeSafely(body: Closure, tn: TestNotifier) {
    try {
      body.run()
    } catch {
      case e => tn.fireFailure(e)
    }
  }

  private def exitSpec() {
    current = current.parent
  }

  def shareSideEffects() {
    current.shareSideEffects()
  }

  def defer(body: => Unit) {
    current.addDefer(new Closure {
      def run() {
        body
      }
    })
  }

  def postponedPaths: List[Path] = {
    assertStatusIs(Finished)
    postponed.toList
  }

  private def changeStatus(from: Status, to: Status) {
    assertStatusIs(from)
    status = to
  }

  private def assertStatusIs(expected: Status) {
    assert(status == expected, "expected status " + expected + ", but was " + status)
  }
}

File: specsy-core/src/main/java/org/specsy/core/Context.java
// Copyright © 2010-2016, Esko Luontola <www.orfjackal.net>
// This software is released under the Apache License 2.0.
// The license text is at http://www.apache.org/licenses/LICENSE-2.0

package org.specsy.core;

import fi.jumi.api.drivers.SuiteNotifier;
import fi.jumi.api.drivers.TestNotifier;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import static org.specsy.core.Context.Status.*;

public class Context {

    enum Status {
        NOT_STARTED,
        RUNNING,
        FINISHED
    }

    private final Path targetPath;
    private final SuiteNotifier notifier;

    private Status status = NOT_STARTED;
    private SpecContext current = null;
    private final List<Path> postponed = new ArrayList<>();

    public Context(Path targetPath, SuiteNotifier notifier) {
        this.targetPath = targetPath;
        this.notifier = notifier;
    }


    // control flow

    public void bootstrap(String className, Closure rootSpec) {
        assertNonNull("className", className);
        assertNonNull("rootSpec", rootSpec);
        changeStatus(NOT_STARTED, RUNNING);

        enterRootSpec(className);
        processSpec(rootSpec);
        exitSpec();

        changeStatus(RUNNING, FINISHED);
    }

    private void enterRootSpec(String name) {
        current = new SpecContext(name, null, Path.ROOT, targetPath);
    }

    public void spec(String name, Closure body) {
        assertNonNull("name", name);
        assertNonNull("body", body);
        assertStatusIs(RUNNING);

        enterSpec(name);
        processSpec(body);
        exitSpec();
    }

    private void enterSpec(String name) {
        current = current.nextChild(name);
    }

    private void processSpec(Closure spec) {
        if (current.shouldExecute()) {
            execute(spec);
        }
        if (current.shouldPostpone()) {
            postponed.add(current.path());
        }
    }

    private void exitSpec() {
        current = current.parent();
    }


    // executing

    private void execute(Closure spec) {
        notifier.fireTestFound(current.path().toTestId(), current.name());
        TestNotifier tn = notifier.fireTestStarted(current.path().toTestId());

        executeSafely(spec, tn);
        for (Closure deferBlock : current.deferred()) {
            executeSafely(deferBlock, tn);
        }

        tn.fireTestFinished();
    }

    private void executeSafely(Closure closure, TestNotifier tn) {
        try {
            closure.run();
        } catch (Throwable t) {
            tn.fireFailure(t);
        }
    }


    // deferring

    public void defer(Closure block) {
        assertNonNull("block", block);
        current.addDefer(block);
    }

    public Iterable<Path> postponedPaths() {
        assertStatusIs(FINISHED);
        return Collections.unmodifiableCollection(postponed);
    }


    // sharing side-effects

    public void shareSideEffects() {
        current.shareSideEffects();
    }


    // status

    private void changeStatus(Status from, Status to) {
        assertStatusIs(from);
        status = to;
    }

    private void assertStatusIs(Status expected) {
        if (status != expected) {
            throw new IllegalStateException("expected status " + expected + ", but was " + status);
        }
    }

    private static void assertNonNull(String varName, Object var) {
        if (var == null) {
            throw new NullPointerException(varName);
        }
    }
}
